File: src/main/java/rag/config/ConfigLoader.java
package main.java.rag.config;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ConfigLoader {
    private final Path configPath;

    public ConfigLoader(Path configPath) {
        this.configPath = configPath;
    }

    public Config loadConfig() {
        try {
            List<String> lines = Files.readAllLines(configPath);
            Map<String, String> configMap = parseYaml(lines);
            
            // Extract values from the parsed map
            String intentType = configMap.get("pipeline.intent_detector");
            String writerType = configMap.get("pipeline.query_writer");
            String retrieverType = configMap.get("pipeline.retriever");
            String reranker = configMap.get("pipeline.reranker");
            String answerAgentType = configMap.get("pipeline.answer_agent");
            
            String rulesFile = configMap.get("params.intent.rules_file");
            int topK = Integer.parseInt(configMap.get("params.retriever.top_k"));
            String stopwordsFile = configMap.get("params.query_writer.stopwords_file");
            int topN = Integer.parseInt(configMap.get("params.query_writer.top_n"));
            
            String documentsDir = configMap.get("paths.documents_dir");
            String chunkStore = configMap.get("paths.chunk_store");
            String logsDir = configMap.get("paths.logs_dir");
            
            // Convert relative paths to absolute paths based on config file location
            Path baseDir = configPath.getParent().getParent(); // Go up from resources to main
            Path rulesFilePath = resolvePath(baseDir, rulesFile);
            Path stopwordsFilePath = resolvePath(baseDir, stopwordsFile);
            Path documentsDirPath = resolvePath(baseDir, documentsDir);
            Path chunkPath = resolvePath(baseDir, chunkStore);
            Path logsDirPath = resolvePath(baseDir, logsDir);
            
            return new Config(
                intentType,
                writerType,
                retrieverType,
                reranker,
                answerAgentType,
                rulesFilePath,
                topK,
                stopwordsFilePath,
                topN,
                documentsDirPath,
                chunkPath,
                logsDirPath
            );
        } catch (IOException e) {
            throw new RuntimeException("Failed to load configuration from: " + configPath, e);
        }
    }
    
    private Map<String, String> parseYaml(List<String> lines) {
        Map<String, String> configMap = new HashMap<>();
        String currentSection = "";
        String currentSubSection = "";
        
        for (String line : lines) {
            if (line.trim().isEmpty() || line.trim().startsWith("#")) {
                continue;
            }
            
            int indentLevel = getIndentLevel(line);
            String content = line.trim();
            
            if (content.endsWith(":")) {
                // Section header
                String section = content.substring(0, content.length() - 1);
                if (indentLevel == 0) {
                    // Top-level section
                    currentSection = section;
                    currentSubSection = "";
                } else if (indentLevel == 2) {
                    // Sub-section under params
                    if (currentSection.equals("params")) {
                        currentSubSection = currentSection + "." + section;
                    } else {
                        currentSubSection = currentSection + "." + section;
                    }
                }
            } else if (content.contains(":")) {
                // Key-value pair
                String[] parts = content.split(":", 2);
                String key = parts[0].trim();
                String value = parts[1].trim();
                
                // Remove quotes if present
                if (value.startsWith("\"") && value.endsWith("\"")) {
                    value = value.substring(1, value.length() - 1);
                }
                
                // Build full key path
                String fullKey;
                if (indentLevel == 2 && !currentSubSection.isEmpty()) {
                    // Nested value (e.g., params.intent.rules_file)
                    fullKey = currentSubSection + "." + key;
                } else if (indentLevel == 2) {
                    // Direct value under section (e.g., pipeline.intent_detector)
                    fullKey = currentSection + "." + key;
                } else if (indentLevel == 0) {
                    // Top-level value (shouldn't happen in our YAML, but handle it)
                    fullKey = key;
                } else {
                    // Default: use current section
                    fullKey = currentSection.isEmpty() ? key : currentSection + "." + key;
                }
                
                configMap.put(fullKey, value);
            }
        }
        
        return configMap;
    }
    
    private int getIndentLevel(String line) {
        int count = 0;
        for (char c : line.toCharArray()) {
            if (c == ' ') {
                count++;
            } else {
                break;
            }
        }
        return count;
    }
    
    private Path resolvePath(Path baseDir, String relativePath) {
        // Remove leading "./" if present
        if (relativePath.startsWith("./")) {
            relativePath = relativePath.substring(2);
        }
        
        // Handle "../" for going up directories
        Path resolved = baseDir;
        while (relativePath.startsWith("../")) {
            resolved = resolved.getParent();
            relativePath = relativePath.substring(3);
        }
        
        return resolved.resolve(relativePath).normalize();
    }
}

File: iteration1_py/src/config/config_loader.py
import yaml
from pathlib import Path
from typing import Dict, Optional

from src.config.config import Config


class ConfigLoader:
    """
    Loads configuration from YAML file.
    Parses the config.yaml file and creates a Config object.
    """
    
    def __init__(self, config_path: Path):
        """
        Initialize ConfigLoader.
        
        Args:
            config_path: Path to the config.yaml file
        """
        self.config_path = Path(config_path)
    
    def load_config(self) -> Config:
        """
        Load and parse configuration from YAML file.
        
        Returns:
            Config object with parsed settings
            
        Raises:
            FileNotFoundError: If config file doesn't exist
            ValueError: If config is malformed
        """
        if not self.config_path.exists():
            raise FileNotFoundError(f"Config file not found: {self.config_path}")
        
        try:
            with open(self.config_path, 'r', encoding='utf-8') as f:
                config_data = yaml.safe_load(f)
            
            if not config_data:
                raise ValueError("Config file is empty")
            
            # Extract pipeline configuration
            pipeline = config_data.get('pipeline', {})
            intent_type = pipeline.get('intent_detector', 'RuleIntentDetector')
            writer_type = pipeline.get('query_writer', 'HeuristicQueryWriter')
            retriever_type = pipeline.get('retriever', 'KeywordRetriever')
            reranker_type = pipeline.get('reranker', 'SimpleReranker')
            answer_agent_type = pipeline.get('answer_agent', 'TemplateAnswerAgent')
            
            # Extract parameters
            params = config_data.get('params', {})
            
            intent_params = params.get('intent', {})
            rules_file = intent_params.get('rules_file', 'resources/intent_rules.yaml')
            
            retriever_params = params.get('retriever', {})
            top_k = retriever_params.get('top_k', 5)
            
            writer_params = params.get('query_writer', {})
            stopwords_file = writer_params.get('stopwords_file', 'resources/stopwords.yaml')
            top_n = writer_params.get('top_n', 3)
            
            # Extract paths
            paths = config_data.get('paths', {})
            chunk_store = paths.get('chunk_store', 'data/chunks.json')
            logs_dir = paths.get('logs_dir', 'logs')
            
            # Resolve paths relative to config file directory
            base_dir = self.config_path.parent
            rules_file_path = self._resolve_path(base_dir, rules_file)
            stopwords_file_path = self._resolve_path(base_dir, stopwords_file)
            chunk_path = self._resolve_path(base_dir, chunk_store)
            logs_dir_path = self._resolve_path(base_dir, logs_dir)
            
            return Config(
                intent_type=intent_type,
                writer_type=writer_type,
                retriever_type=retriever_type,
                reranker_type=reranker_type,
                answer_agent_type=answer_agent_type,
                rules_file_path=rules_file_path,
                top_k=int(top_k),
                stopwords_file_path=stopwords_file_path,
                top_n=int(top_n),
                chunk_path=chunk_path,
                logs_dir_path=logs_dir_path
            )
        
        except yaml.YAMLError as e:
            raise ValueError(f"Failed to parse YAML config: {e}")
        except Exception as e:
            raise RuntimeError(f"Failed to load configuration from {self.config_path}: {e}")
    
    @staticmethod
    def _resolve_path(base_dir: Path, relative_path: str) -> Path:
        """
        Resolve a path relative to base directory.
        If path is absolute, return as-is.
        
        Args:
            base_dir: Base directory for relative paths
            relative_path: Path to resolve
            
        Returns:
            Resolved absolute Path
        """
        path = Path(relative_path)
        
        if path.is_absolute():
            return path
        
        return (base_dir / path).resolve()
